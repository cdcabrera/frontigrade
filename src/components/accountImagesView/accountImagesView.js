import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, Breadcrumb, EmptyState, Grid, ListView, Modal, Row, Spinner } from 'patternfly-react';
import { withRouter } from 'react-router-dom';
import _isEqual from 'lodash/isEqual';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';
import apiTypes from '../../constants/apiConstants';
import accountImagesViewTypes from './accountImagesViewConstants';
import ViewToolbar from '../viewToolbar/viewToolbar';
import AccountImagesViewInstanceGraphs from './accountImagesViewInstanceGraphs';
import AccountImagesViewListItem from './accountImagesViewListItem';

class AccountImagesView extends React.Component {
  state = {
    accountId: null
  };

  componentDidMount() {
    this.loadApi();
  }

  componentDidUpdate(prevProps) {
    const { filter, match } = this.props;

    if (
      match.params.accountId !== prevProps.match.params.accountId ||
      !_isEqual(filter.query, prevProps.filter.query)
    ) {
      this.loadApi();
    }
  }

  onAddAccount = () => {
    this.backToAccounts();

    store.dispatch({
      type: reduxTypes.account.ADD_ACCOUNT_SHOW
    });
  };

  /**
   * FixMe: API - issue
   * The API requires the account id be passed as a query param, counter to other account API calls.
   * In order to emulate consistent behavior we handle that query bundle at the service level instead,
   * see services/accountServices.js "getAccountImages"
   */
  loadApi() {
    const { filter, getAccount, getAccountImages, match } = this.props;
    const accountId = match && Number.parseInt(match.params && match.params.accountId, 10);

    if (!Number.isNaN(accountId)) {
      this.setState({ accountId }, () => {
        getAccount(accountId);

        if (filter.query) {
          getAccountImages(accountId, filter.query);
        }
      });
    } else {
      this.backToAccounts();
    }
  }

  backToAccounts(event = {}) {
    const { history } = this.props;

    if (event.preventDefault) {
      event.preventDefault();
    }

    history.push('/accounts/');
  }

  renderList() {
    const { images } = this.props;

    if (images.length) {
      return (
        <ListView className="cloudmeter-list-view">
          {images.map(item => (
            <AccountImagesViewListItem
              item={item}
              key={item[apiTypes.API_RESPONSE_IMAGES_ID]}
              onDetail={this.onDetailView}
              onEdit={this.onEditName}
              onArchive={this.onArchive}
            />
          ))}
        </ListView>
      );
    }

    return (
      <EmptyState className="list-view-blank-slate">
        <EmptyState.Title>No Results Match the Filter Criteria</EmptyState.Title>
        <EmptyState.Info>The active filters are hiding all items.</EmptyState.Info>
        <EmptyState.Action>
          <Button bsStyle="link" onClick={this.onClearFilters}>
            Clear Filters
          </Button>
        </EmptyState.Action>
      </EmptyState>
    );
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Modal bsSize="lg" backdrop={false} show animation={false}>
          <Modal.Body>
            <Spinner loading size="lg" className="blank-slate-pf-icon" />
            <div className="text-center">Loading...</div>
          </Modal.Body>
        </Modal>
      );
    }

    return null;
  }

  render() {
    const { account, images, error, errorMessage, filter, pending, view, viewGlobal } = this.props;
    const { accountId } = this.state;

    if (error) {
      return (
        <EmptyState>
          <Alert type="error">
            <span>Error retrieving images: {errorMessage}</span>
          </Alert>
          {this.renderPendingMessage()}
        </EmptyState>
      );
    }

    if (pending && !images.length) {
      return <div className="cloudmeter-view-container">{this.renderPendingMessage()}</div>;
    }

    // ToDo: Replace filterFields={[]} with filterFields={accountImagesViewTypes.filterFields} when the name filter is active.
    if (images.length || filter.activeFilters.length) {
      return (
        <div className="cloudmeter-view-container fadein">
          <Grid fluid>
            <Grid.Row>
              <Grid.Col xs={12}>
                <Breadcrumb title className="cloudmeter-breadcrumb">
                  <Breadcrumb.Item onClick={e => this.backToAccounts(e)}>Accounts</Breadcrumb.Item>
                  <Breadcrumb.Item active aria-current="page">
                    <strong>
                      {account[apiTypes.API_RESPONSE_ACCOUNT_NAME] ||
                        account[apiTypes.API_RESPONSE_ACCOUNT_ACCOUNT_ID] ||
                        `Account ${account[apiTypes.API_RESPONSE_ACCOUNT_ID] || ''}`}
                    </strong>
                  </Breadcrumb.Item>
                </Breadcrumb>
              </Grid.Col>
            </Grid.Row>
          </Grid>
          <ViewToolbar
            dateFields={accountImagesViewTypes.dateFields.timeValues}
            filterFields={[]}
            onAddAccount={this.onAddAccount}
            onExport={null}
            sortFields={accountImagesViewTypes.sortFields}
            view={view}
            viewGlobal={viewGlobal}
            {...filter}
          />
          <div className="cloudmeter-list-container">
            {/\d/.test(accountId) && <AccountImagesViewInstanceGraphs filterId={accountId} filter={filter} />}
            {this.renderList()}
          </div>
          {this.renderPendingMessage()}
        </div>
      );
    }

    return (
      <Grid fluid>
        <Row>
          <EmptyState className="full-page-blank-slate">
            <EmptyState.Icon />
            <EmptyState.Title>No Images Available</EmptyState.Title>
            <EmptyState.Info>Add an AWS account to monitor usage.</EmptyState.Info>
            <EmptyState.Action>
              <Button bsStyle="primary" bsSize="large" onClick={this.onAddAccount}>
                Add Account
              </Button>
            </EmptyState.Action>
          </EmptyState>
        </Row>
      </Grid>
    );
  }
}

AccountImagesView.propTypes = {
  account: PropTypes.object,
  images: PropTypes.array,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  filter: PropTypes.shape({
    activeFilters: PropTypes.array,
    query: PropTypes.object
  }),
  getAccount: PropTypes.func,
  getAccountImages: PropTypes.func,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.object
  }),
  pending: PropTypes.bool,
  view: PropTypes.string,
  viewGlobal: PropTypes.string
};

AccountImagesView.defaultProps = {
  account: {},
  images: [],
  error: false,
  errorMessage: null,
  filter: {
    activeFilters: [],
    query: {}
  },
  getAccount: helpers.noop,
  getAccountImages: helpers.noop,
  match: {
    params: {}
  },
  pending: false,
  view: 'accountImages',
  viewGlobal: 'accountGlobal'
};

const mapDispatchToProps = dispatch => ({
  getAccount: id => dispatch(reduxActions.account.getAccount(id)),
  getAccountImages: (id, query) => dispatch(reduxActions.account.getAccountImages(id, query))
});

const mapStateToProps = state => ({
  ...state.accountImages.view,
  filter: {
    ...state.filter.accountImages,
    ...state.filter.accountGlobal,
    ...{
      query: {
        ...state.filter.accountImages.query,
        ...state.filter.accountGlobal.query
      }
    }
  }
});

const ConnectedAccountImagesView = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(AccountImagesView)
);

export { ConnectedAccountImagesView as default, ConnectedAccountImagesView, AccountImagesView };
