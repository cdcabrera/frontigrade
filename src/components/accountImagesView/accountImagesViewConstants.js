import helpers from '../../common/helpers';
import apiTypes from '../../constants/apiConstants';

const dateFieldTypes = helpers.generatePriorYearMonthArray();

const filterFieldTypes = [
  {
    id: apiTypes.API_QUERY_NAME,
    title: 'Name',
    placeholder: 'Filter by Name',
    filterType: 'text'
  }
];

const sortFieldTypes = [
  {
    id: 'name',
    title: 'Name',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'images',
    title: 'Images',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'instances',
    title: 'Instances',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'rhel',
    title: 'Rhel',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'openshift',
    title: 'OpenShift',
    isNumeric: false,
    sortAscending: true
  }
];

const accountImagesViewTypes = {
  dateFields: dateFieldTypes,
  filterFields: filterFieldTypes,
  sortFields: sortFieldTypes
};

export { accountImagesViewTypes as default, accountImagesViewTypes, dateFieldTypes, filterFieldTypes, sortFieldTypes };
