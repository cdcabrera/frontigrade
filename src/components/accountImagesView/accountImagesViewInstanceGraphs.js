import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardGrid, Label as PFLabel, Spinner, UtilizationCard } from 'patternfly-react';
import _sumBy from 'lodash/sumBy';
import { connect, reduxActions } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';

class AccountImagesViewInstanceGraphs extends React.Component {
  static calculateInstanceTotals(dailyUsage = []) {
    const openshift = _sumBy(dailyUsage, val =>
      Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_RUNTIME])
    );

    const rhel = _sumBy(dailyUsage, val => Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_RUNTIME]));

    return {
      openshiftHours: helpers.generateHoursFromSeconds(openshift),
      rhelHours: helpers.generateHoursFromSeconds(rhel)
    };
  }

  componentDidMount() {
    const { filter, filterId, getAccountImagesInstances } = this.props;

    if (!Number.isNaN(Number.parseInt(filterId, 10)) && filter.query) {
      getAccountImagesInstances(filterId, filter.query);
    }
  }

  componentDidUpdate(prevProps) {
    const { filter, filterId, getAccountImagesInstances, updateInstances } = this.props;

    if (updateInstances === true && updateInstances !== prevProps.updateInstances) {
      getAccountImagesInstances(filterId, filter.query);
    }
  }

  renderPendingCard() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Card matchHeight className="cloudmeter-utilization-graph cloudmeter-utilization-graph-loading fadein">
          <Card.Body>
            <Spinner loading size="sm" className="blank-slate-pf-icon" />
            <div className="text-center">
              <small>Loading...</small>
            </div>
          </Card.Body>
        </Card>
      );
    }

    return null;
  }

  render() {
    const { error, fulfilled, dailyUsage, instancesOpenshift, instancesRhel } = this.props;
    let displayTotals = {};

    if (error) {
      return null;
    }

    if (dailyUsage && dailyUsage.length) {
      displayTotals = { ...AccountImagesViewInstanceGraphs.calculateInstanceTotals(dailyUsage) };
    }

    return (
      <CardGrid matchHeight>
        <CardGrid.Row>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Title>Red Hat Enterprise Linux</Card.Title>
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{instancesRhel}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      <PFLabel bsStyle="warning">
                        <abbr title="Red Hat Enterprise Linux">RHEL</abbr>
                      </PFLabel>{' '}
                      Instances
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Title>Red Hat OpenShift Container Platform</Card.Title>
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{instancesOpenshift}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      <PFLabel bsStyle="primary">
                        <abbr title="Red Hat OpenShift Container Platform">RHOCP</abbr>
                      </PFLabel>{' '}
                      Instances
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{displayTotals.rhelHours || 0}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>Red Hat Enterprise Linux Hours</UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{displayTotals.openshiftHours || 0}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      Red Hat OpenShift Container Platform Hours
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
        </CardGrid.Row>
      </CardGrid>
    );
  }
}

AccountImagesViewInstanceGraphs.propTypes = {
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  filter: PropTypes.shape({
    query: PropTypes.object.isRequired
  }).isRequired,
  filterId: PropTypes.number.isRequired,
  getAccountImagesInstances: PropTypes.func,
  dailyUsage: PropTypes.array,
  instancesOpenshift: PropTypes.number,
  instancesRhel: PropTypes.number,
  pending: PropTypes.bool,
  updateInstances: PropTypes.bool
};

AccountImagesViewInstanceGraphs.defaultProps = {
  error: false,
  fulfilled: false,
  getAccountImagesInstances: helpers.noop,
  dailyUsage: [],
  instancesOpenshift: 0,
  instancesRhel: 0,
  pending: false,
  updateInstances: false
};

const mapDispatchToProps = dispatch => ({
  getAccountImagesInstances: (id, query) => dispatch(reduxActions.account.getAccountImagesInstances(id, query))
});

const mapStateToProps = state => ({
  ...state.accountImages.instances
});

const ConnectedAccountImagesViewInstanceGraphs = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountImagesViewInstanceGraphs);

export {
  ConnectedAccountImagesViewInstanceGraphs as default,
  ConnectedAccountImagesViewInstanceGraphs,
  AccountImagesViewInstanceGraphs
};
