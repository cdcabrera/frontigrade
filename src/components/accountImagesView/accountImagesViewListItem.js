import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Icon, ListView, Label as PFLabel } from 'patternfly-react';
import { connectTranslate } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';
import AccountImagesViewListItemDetail from './accountImagesViewListItemDetail';
import Tooltip from '../tooltip/tooltip';

class AccountImagesViewListItem extends React.Component {
  static renderLeftContent() {
    return <ListView.Icon name="unknown" />;
  }

  rhelFilteredSecondsHours() {
    const { filter, item } = this.props;

    let rhelSecondsHours;

    switch (filter.graphRhelValue) {
      case 'rhelMemoryTime':
        rhelSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_MEMORY];
        break;
      case 'rhelVcpuTime':
        rhelSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_VCPU];
        break;
      default:
        rhelSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_RUNTIME];
        break;
    }

    return helpers.generateHoursFromSeconds(rhelSecondsHours);
  }

  rhocpFilteredSecondsHours() {
    const { filter, item } = this.props;

    let rhocpSecondsHours;

    switch (filter.graphOpenshiftValue) {
      case 'openshiftMemoryTime':
        rhocpSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_MEMORY];
        break;
      case 'openshiftVcpuTime':
        rhocpSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_VCPU];
        break;
      default:
        rhocpSecondsHours = item[apiTypes.API_RESPONSE_IMAGES_RUNTIME];
        break;
    }

    return helpers.generateHoursFromSeconds(rhocpSecondsHours);
  }

  renderHeading() {
    const { item } = this.props;

    const timestamp = item[apiTypes.API_RESPONSE_IMAGES_EDIT_UPDATED];
    const title =
      item[apiTypes.API_RESPONSE_IMAGES_NAME] ||
      item[apiTypes.API_RESPONSE_IMAGES_IMAGE_ID] ||
      `Image #${item[apiTypes.API_RESPONSE_IMAGES_ID] || ''}`;

    const titleOverflow = title && title.length > 25 ? { title } : {};

    return (
      <span className="cloudmeter-list-view-item-heading">
        <strong className="cloudmeter-list-view-item-heading-title">
          <span {...titleOverflow}>{title}</span>
        </strong>
        {timestamp && (
          <React.Fragment>
            <br />
            Updated {moment(timestamp).format('h:mmA, MMMM Do YYYY')}
          </React.Fragment>
        )}
      </span>
    );
  }

  static renderDescription() {
    return null;
  }

  /**
   * FixMe: PF-React issue
   * listview "additionalInfo" attribute "requires" an array propType, restrictive limit.
   * Open it up to allow both, "node" OR "array"
   */
  renderAdditionalInfo() {
    const { item, t } = this.props;

    const rhelChallenged = item[apiTypes.API_RESPONSE_IMAGES_RHEL_CHALLENGED];
    const openshiftChallenged = item[apiTypes.API_RESPONSE_IMAGES_OPENSHIFT_CHALLENGED];

    const rhelSecondsHours = !item[apiTypes.API_RESPONSE_IMAGES_RHEL] ? null : this.rhelFilteredSecondsHours();
    const rhocpSecondsHours = !item[apiTypes.API_RESPONSE_IMAGES_OPENSHIFT] ? null : this.rhocpFilteredSecondsHours();

    let instances = Number.parseInt(item[apiTypes.API_RESPONSE_IMAGES_INSTANCES], 10);
    instances = Number.isNaN(instances) ? 'N/A' : instances;

    const instancesPopover = t('list-images.instances.instances-tooltip');
    const rhelPopover = t('list-images.rhel.rhel-tooltip');
    const rhocpPopover = t('list-images.rhocp.rhocp-tooltip');

    return [
      <ListView.InfoItem key="1" className="cloudmeter-listview-infoitem">
        <Tooltip delayShow={100} popover={instancesPopover} trigger="click">
          <Icon type="pf" name="screen" />
          <strong>{instances}</strong> Instances
        </Tooltip>
      </ListView.InfoItem>,
      <ListView.InfoItem key="2" className="cloudmeter-listview-label cloudmeter-listview-label-has-badge">
        {rhelSecondsHours !== null && (
          <Tooltip tooltip={`${rhelSecondsHours.seconds} seconds`} placement="bottom">
            <Icon type="fa" name="clock-o" />
            <strong data-test="rhel">{rhelSecondsHours.hours}</strong>
          </Tooltip>
        )}
        {rhelSecondsHours === null && (
          <React.Fragment>
            <Icon type="fa" name="clock-o" />
            <strong data-test="rhel">N/A</strong>
          </React.Fragment>
        )}
        <Tooltip delayShow={100} popover={rhelPopover} trigger="click">
          <PFLabel bsStyle="warning">
            <abbr title="Red Hat Enterprise Linux">RHEL</abbr>
          </PFLabel>{' '}
          {rhelChallenged && <Icon type="fa" name="flag" className="cloudmeter-pficon-error" />}
        </Tooltip>
      </ListView.InfoItem>,
      <ListView.InfoItem key="3" className="cloudmeter-listview-label cloudmeter-listview-label-has-badge">
        {rhocpSecondsHours !== null && (
          <Tooltip tooltip={`${rhocpSecondsHours.seconds} seconds`} placement="bottom">
            <Icon type="fa" name="clock-o" />
            <strong data-test="rhocp">{rhocpSecondsHours.hours}</strong>
          </Tooltip>
        )}
        {rhocpSecondsHours === null && (
          <React.Fragment>
            <Icon type="fa" name="clock-o" />
            <strong data-test="rhocp">N/A</strong>
          </React.Fragment>
        )}
        <Tooltip delayShow={100} popover={rhocpPopover} trigger="click">
          <PFLabel bsStyle="primary">
            <abbr title="Red Hat OpenShift Container Platform">RHOCP</abbr>
          </PFLabel>{' '}
          {openshiftChallenged && <Icon type="fa" name="flag" className="cloudmeter-pficon-error" />}
        </Tooltip>
      </ListView.InfoItem>
    ];
  }

  render() {
    const { item } = this.props;
    const itemId = item[apiTypes.API_RESPONSE_IMAGES_ID];

    return (
      <ListView.Item
        className="cloudmeter-accountview-list-view-item"
        key={itemId}
        leftContent={AccountImagesViewListItem.renderLeftContent()}
        heading={this.renderHeading()}
        description={AccountImagesViewListItem.renderDescription()}
        additionalInfo={this.renderAdditionalInfo()}
        stacked={false}
        aria-live="polite"
      >
        <AccountImagesViewListItemDetail id={itemId} />
      </ListView.Item>
    );
  }
}

AccountImagesViewListItem.propTypes = {
  filter: PropTypes.shape({
    graphOpenshiftValue: PropTypes.string,
    graphRhelValue: PropTypes.string
  }),
  item: PropTypes.object.isRequired,
  t: PropTypes.func
};

AccountImagesViewListItem.defaultProps = {
  filter: {
    graphOpenshiftValue: null,
    graphRhelValue: null
  },
  t: helpers.noopTranslate
};

const ConnectedAccountImagesViewListItem = connectTranslate()(AccountImagesViewListItem);

export { ConnectedAccountImagesViewListItem as default, ConnectedAccountImagesViewListItem, AccountImagesViewListItem };
