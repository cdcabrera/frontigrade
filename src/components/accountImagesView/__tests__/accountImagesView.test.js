import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import helpers from '../../../common/helpers';
import { ConnectedAccountImagesView, AccountImagesView } from '../accountImagesView';

describe('AccountImagesView Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component with default props', () => {
    const store = generateEmptyStore({ account: {}, filter: { account: {} } });
    const component = shallow(<ConnectedAccountImagesView />, { context: { store } });

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a non-connected component with empty state', () => {
    const props = {
      history: {
        push: helpers.noop
      }
    };

    const component = mount(<AccountImagesView {...props} />);

    expect(component.find('button').length).toEqual(1);
    expect(component.render()).toMatchSnapshot('empty state');
  });

  it('should render a non-connected component in pending state', () => {
    const props = {
      pending: true,
      history: {
        push: helpers.noop
      }
    };

    const component = shallow(<AccountImagesView {...props} />);

    expect(component.render()).toMatchSnapshot('pending');

    component.setProps({
      images: [
        {
          cloud_image_id: 'ami-rhel7',
          id: 2,
          instances_seen: 2,
          is_encrypted: false,
          name: null,
          openshift: false,
          openshift_challenged: false,
          openshift_detected: false,
          rhel: true,
          rhel_challenged: false,
          rhel_detected: true,
          runtime_seconds: 90362.5,
          status: 'inspected'
        }
      ],
      filter: {
        dateFields: null
      }
    });

    expect(component).toMatchSnapshot('pending with list');
  });

  it('should render a non-connected component with a list', () => {
    const props = {
      images: [
        {
          cloud_image_id: 'ami-rhel7',
          id: 2,
          instances_seen: 2,
          is_encrypted: false,
          name: null,
          openshift: false,
          openshift_challenged: false,
          openshift_detected: false,
          rhel: true,
          rhel_challenged: false,
          rhel_detected: true,
          runtime_seconds: 90362.5,
          status: 'inspected'
        }
      ],
      filter: {
        dateFields: null
      },
      history: {
        push: helpers.noop
      }
    };

    const component = mount(<AccountImagesView {...props} />);

    expect(component.find('.cloudmeter-list-container').length).toEqual(1);
    expect(component.find('.cloudmeter-list-container').render()).toMatchSnapshot('list');
  });
});
