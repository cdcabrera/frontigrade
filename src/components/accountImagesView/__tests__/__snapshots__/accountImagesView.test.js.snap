// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountImagesView Component should render a connected component with default props: connected 1`] = `
<Route>
  <Component />
</Route>
`;

exports[`AccountImagesView Component should render a non-connected component in fulfilled state: fulfilled and empty 1`] = `
<div
  class="container-fluid"
>
  <div
    class="row"
  >
    <div
      class="blank-slate-pf full-page-blank-slate"
    >
      <div
        class="blank-slate-pf-icon"
      >
        <span
          aria-hidden="true"
          class="pficon pficon-add-circle-o"
        />
      </div>
      <h4
        class="h1 blank-slate-pf-title"
      >
        No Images Available
      </h4>
      <p
        class="blank-slate-pf-info"
      >
        Add an AWS account to monitor usage.
      </p>
      <div
        class="blank-slate-pf-main-action"
      >
        <button
          class="btn btn-lg btn-primary"
          type="button"
        >
          Add Account
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`AccountImagesView Component should render a non-connected component in fulfilled state: fulfilled with a empty list 1`] = `
<div
  className="cloudmeter-view-container fadein"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={true}
  >
    <Breadcrumb
      className="cloudmeter-breadcrumb"
      title={true}
    >
      <BreadcrumbItem
        active={false}
        onClick={[Function]}
      >
        Accounts
      </BreadcrumbItem>
      <BreadcrumbItem
        active={true}
        aria-current="page"
      >
        <strong>
          Account 
        </strong>
      </BreadcrumbItem>
    </Breadcrumb>
  </Grid>
  <ViewToolbar
    activeFilters={Array []}
    dateFields={null}
    dateValue={null}
    filterField={null}
    filterFields={Array []}
    filterValue=""
    onAddAccount={[Function]}
    onExport={null}
    query={Object {}}
    resultsType="Result"
    resultsTypePlural="Results"
    selectedCount={0}
    sortAscending={true}
    sortFields={
      Array [
        Object {
          "id": "name",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Name",
        },
        Object {
          "id": "images",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Images",
        },
        Object {
          "id": "instances",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Instances",
        },
        Object {
          "id": "rhel",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Rhel",
        },
        Object {
          "id": "openshift",
          "isNumeric": false,
          "sortAscending": true,
          "title": "OpenShift",
        },
      ]
    }
    sortValue={null}
    totalCount={0}
    view="accountImages"
    viewGlobal="accountGlobal"
  />
  <div
    className="cloudmeter-list-container"
  >
    <EmptyState
      className="list-view-blank-slate"
    >
      <EmptyStateTitle
        className=""
      >
        No Results Available
      </EmptyStateTitle>
      <EmptyStateAction
        className=""
        secondary={false}
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="link"
          disabled={false}
          onClick={[Function]}
        >
          Reset Date Selector
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="link"
          disabled={false}
          onClick={[Function]}
        >
          Return to Accounts
        </Button>
      </EmptyStateAction>
    </EmptyState>
  </div>
</div>
`;

exports[`AccountImagesView Component should render a non-connected component in fulfilled state: fulfilled with a list 1`] = `
<div
  className="cloudmeter-view-container fadein"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={true}
  >
    <Breadcrumb
      className="cloudmeter-breadcrumb"
      title={true}
    >
      <BreadcrumbItem
        active={false}
        onClick={[Function]}
      >
        Accounts
      </BreadcrumbItem>
      <BreadcrumbItem
        active={true}
        aria-current="page"
      >
        <strong>
          Account 
        </strong>
      </BreadcrumbItem>
    </Breadcrumb>
  </Grid>
  <ViewToolbar
    activeFilters={Array []}
    dateFields={null}
    dateValue={null}
    filterField={null}
    filterFields={Array []}
    filterValue=""
    onAddAccount={[Function]}
    onExport={null}
    query={Object {}}
    resultsType="Result"
    resultsTypePlural="Results"
    selectedCount={0}
    sortAscending={true}
    sortFields={
      Array [
        Object {
          "id": "name",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Name",
        },
        Object {
          "id": "images",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Images",
        },
        Object {
          "id": "instances",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Instances",
        },
        Object {
          "id": "rhel",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Rhel",
        },
        Object {
          "id": "openshift",
          "isNumeric": false,
          "sortAscending": true,
          "title": "OpenShift",
        },
      ]
    }
    sortValue={null}
    totalCount={0}
    view="accountImages"
    viewGlobal="accountGlobal"
  />
  <div
    className="cloudmeter-list-container"
  >
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={true}
    >
      <Row
        bsClass="row"
        className="row-cards-pf"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          xs={12}
        >
          <Card
            accented={false}
            aggregated={false}
            aggregatedMini={false}
            cardRef={null}
            className="cloudmeter-list-view-card"
            matchHeight={false}
          >
            <ListView
              className="cloudmeter-list-view"
            >
              <Connect(Translate(AccountImagesViewListItem))
                filter={
                  Object {
                    "dateFields": null,
                    "query": Object {},
                  }
                }
                item={
                  Object {
                    "cloud_image_id": "ami-rhel7",
                    "id": 2,
                    "instances_seen": 2,
                    "is_encrypted": false,
                    "memory_seconds": 500.5,
                    "name": null,
                    "openshift": true,
                    "openshift_challenged": true,
                    "openshift_detected": false,
                    "rhel": true,
                    "rhel_challenged": false,
                    "rhel_detected": true,
                    "runtime_seconds": 2000.4,
                    "status": "inspected",
                    "vcpu_seconds": 500.2,
                  }
                }
                key="2"
              />
            </ListView>
          </Card>
        </Col>
      </Row>
    </Grid>
  </div>
</div>
`;

exports[`AccountImagesView Component should render a non-connected component in pending state: pending 1`] = `
<div
  class="cloudmeter-view-container"
/>
`;

exports[`AccountImagesView Component should render a non-connected component in pending state: pending with no visible list 1`] = `
<div
  className="cloudmeter-view-container"
>
  <Modal
    animation={false}
    autoFocus={true}
    backdrop={false}
    bsClass="modal"
    bsSize="lg"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <Spinner
        className="blank-slate-pf-icon"
        inline={false}
        inverse={false}
        loading={true}
        size="lg"
      />
      <div
        className="text-center"
      >
        Loading...
      </div>
    </ModalBody>
  </Modal>
</div>
`;

exports[`AccountImagesView Component should render a non-connected component with empty state: empty state 1`] = `
<div
  class="container-fluid"
>
  <div
    class="row"
  >
    <div
      class="blank-slate-pf full-page-blank-slate"
    >
      <div
        class="blank-slate-pf-icon"
      >
        <span
          aria-hidden="true"
          class="pficon pficon-add-circle-o"
        />
      </div>
      <h4
        class="h1 blank-slate-pf-title"
      >
        No Images Available
      </h4>
      <p
        class="blank-slate-pf-info"
      >
        Add an AWS account to monitor usage.
      </p>
      <div
        class="blank-slate-pf-main-action"
      >
        <button
          class="btn btn-lg btn-primary"
          type="button"
        >
          Add Account
        </button>
      </div>
    </div>
  </div>
</div>
`;
